### ISSUE TITLE
update `k8s.io/client-go` release process docs to stop adding major version tags

### ISSUE THREAD

**Issue by liggitt** [kubernetes/kubernetes#84372]
go modules require libraries that are module-aware (contain a go.mod file) and use a semver tag >=2.0 to include the major version as a suffix of the module name (e.g. `example.com/library/v2`).

https://github.com/golang/go/wiki/Modules#semantic-import-versioning
> As a result of Semantic Import Versioning, code opting in to Go modules must comply with these rules:
>
> * If the module is version v2 or higher, the major version of the module must be included as a /vN at the end of the module paths used in go.mod files (e.g., module github.com/my/mod/v2, require github.com/my/mod/v2 v2.0.1) and in the package import path (e.g., import "github.com/my/mod/v2/mypkg"). This includes the paths used in go get commands (e.g., go get github.com/my/mod/v2@v2.0.1. Note there is both a /v2 and a @v2.0.1 in that example. One way to think about it is that the module name now includes the /v2, so include /v2 whenever you are using the module name).

The fact that this was required was unclear previously. go1.13 begins enforcing this and refuses to build a module required at v2.0.0 whose name does not end in `/v2`

The existing `v12.0.0` tag is not a valid module (the `kubernetes-1.15.0` tag is), and go1.13 will refuse to fetch/build it.

with go1.13:
```sh
go get k8s.io/client-go@v12.0.0
go: finding k8s.io/client-go v12.0.0
go: finding k8s.io/client-go v12.0.0
go get k8s.io/client-go@v12.0.0: k8s.io/client-go@v12.0.0: invalid version: module contains a go.mod file, so major version must be compatible: should be v0 or v1, not v12
```

Note that non-semver tags are still addressable via go get (e.g. `k8s.io/client-go@kubernetes-1.15.0`) and require directives (using resolved `v0.0.0-$date-$sha` "pseudoversions").

with go1.13:
```sh
$ go get k8s.io/client-go@kubernetes-1.15.0
go: finding k8s.io kubernetes-1.15.0
go: finding k8s.io/client-go kubernetes-1.15.0
go: downloading k8s.io/client-go v0.0.0-20190620085101-78d2af792bab
go: extracting k8s.io/client-go v0.0.0-20190620085101-78d2af792bab
```

/sig api-machinery
cc @sttts @nikhita @BenTheElder

**Comment by liggitt** [comment]
For the v12.0.0 tag, I'm inclined to leave it as-is for git consumers, potentially updating the release note in github to indicate what should be used instead with go modules

We can continue publishing non-semver tags to make new client-go releases addressable.

Before creating a new major version tag, we would need to enable semantic import versioning on the client-go repo.

**Comment by liggitt** [comment]
/kind documentation
/remove-kind bug

this is more of a process/awareness issue, I guess

**Comment by sttts** [comment]
> this is more of a process/awareness issue, I guess

@nikhita do we have docs for the client-go release process anywhere?

**Comment by nikhita** [comment]
Starting in v1.17.0+, for Kubernetes v1.x.y, published repos have corresponding v0.x.y semver tags (https://github.com/kubernetes/kubernetes/issues/84608) and we aren't manually tagging major versions for client-go anymore.

Docs are being added in https://github.com/kubernetes/client-go/pull/721 and https://github.com/kubernetes/kubernetes/pull/85995.

IMO this can be closed now that we don't tag major versions anymore.

**Comment by nikhita** [comment]
We can use https://github.com/kubernetes/kubernetes/issues/72638 to track semantic import versioning for published repos

**Comment by pohly** [comment]
> For the v12.0.0 tag, I'm inclined to leave it as-is for git consumers, potentially updating the release note in github to indicate what should be used instead with go modules

The v12.0.0 and v11.0.0 are still a nuisance when using `go get -u ./...`: as a test, I updated https://github.com/kubernetes-csi/external-attacher/ with `./release-tools/go-get-kubernetes.sh 1.17.0` to 1.17.0, using go 1.13.4. We get `require k8s.io/client-go v0.17.0` in `go.mod` (thanks, much better than before - but see below) and a `replace k8s.io/client-go => k8s.io/client-go v0.17.0` (put in place by that script and resolved to v0.17.0 by Go). But after `go get -u ./...`, the required version of client-go is listed as `v11.0.0+incompatible`.

The `replace` line is unchanged, though, so it still builds with the desired release. It's just confusing. My preference therefore would be to rename those invalid semantic tags to something like `client-go- v12.0.0`. I suspect this would allow us to get rid of the `replace` statements, although having them as a safe-guard might still be wiser.

One other observation from the experiment above: because some repos have multiple tags pointing to the commit that gets picked by `kubernetes-1.17.0` and Go considers all of them as valid choices when not using a stable version (my guess - if I remember correctly there is some logic for when `-beta` tags are used and when they are ignored), we get apparently inconsistent versions in `go.mod`:
```
        k8s.io/api v0.17.0
        k8s.io/apimachinery v0.17.1-beta.0     <===
        k8s.io/client-go v0.17.0
        k8s.io/csi-translation-lib v0.17.0
```

**Comment by liggitt** [comment]
> The v12.0.0 and v11.0.0 are still a nuisance when using go get -u ./...

Removing those major version tags would not make `go get -u` work well with Kubernetes repos until we move to semantic import versioning. Even if those major version tags were removed, `go get -u ...` would still bump from `v0.17.0` to `v0.18.0`, etc.

If you just want to pick up patch versions, `go get -u=patch` might work better for you.

> My preference therefore would be to rename those invalid semantic tags to something like client-go- v12.0.0.

Removing existing tags is *extremely* poor behavior for a repository; I don't anticipate us doing that, especially since it wouldn't actually make `go get -u` work well.

> One other observation from the experiment above: because some repos have multiple tags pointing to the commit that gets picked by kubernetes-1.17.0

We noticed that as well. Our current recommendation is to use semver tags with `go get`. `k8s.io/apimachinery` is the only repo affected by this. We may consider forcing a change in the go.mod file in the future (e.g. a comment like "# Corresponds to Kubernetes v1.17.0") which would cause a different SHA for the v1.17.0 and v1.17.1-beta.0 tags
